Add WebExtensionSidebar object to UIProcess.

AriYoung00 opened the pull request:
#### dd82fa0c239ee137d28c8b225390b0d482d9b401
<pre>
Add WebExtensionSidebar object to UIProcess.
<a href="https://webkit.org/b/277574">https://webkit.org/b/277574</a>
<a href="https://rdar.apple.com/133073139">rdar://133073139</a>

Reviewed by Timothy Hatcher.

This PR adds the WebExtensionSidebar class to UIProcess. The structure
of this class closely mirrors that of WebExtensionAction, except it
prefers to use std::optional over nullable pointers when possible. This
PR also adds several utility methods in WebExtensionContext which will
be used to manage the WebExtensionSidebar objects necessary to service
the sidePanel / sidebarAction APIs for each extensions that utilizes
them.

* Source/WebKit/Platform/cocoa/CocoaHelpers.h:
 (WebKit::toOptional): Add helper method to convert nulltable T * (where
T is some Cocoa type) to optional&lt;RetainPtr&lt;T&gt;&gt;, where the interior
RetainPtr is guaranteed to be non-null
* Source/WebKit/Shared/API/APIObject.h: Add WebExtensionSidebar to
  API::Object::Type enum
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::sidebarIcon): Getter for current extension&apos;s
default sidebar icon
(WebKit::WebExtension::sidebarDocumentPath): Getter for current
extension&apos;s default document path
(WebKit::WebExtension::sidebarTitle): Getter for current extension&apos;s
default sidebar title
(WebKit::WebExtension::populateSidebarPropertiesIfNeeded): Helper method
to populate sidebar-related properties if needed
(WebKit::WebExtension::populateSidebarActionProperties): Helper method
to populate sidebar-related properties when sidebarAction is detected
(WebKit::WebExtension::populateSidePanelProperties): Helper method to
populate sidebar-related properties when sidePanel is detected
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::unload): Set m_defaultSidebar to nullptr
on unload
(WebKit::WebExtensionContext::defaultSidebar): Getter for
m_defaultSidebar
(WebKit::WebExtensionContext::getSidebar): Getters for window- or
tab-specific sidebar objects
(WebKit::WebExtensionContext::getOrCreateSidebar): Functions to create
and get window- or tab-specific sidebar objects
* Source/WebKit/UIProcess/Extensions/WebExtensionSidebar.h: Added.
(WebKit::WebExtensionSidebar::create): Canonical WebKit create method
(WebKit::WebExtensionSidebar::isOpen const): Getter for m_isOpen
(WebKit::WebExtensionSidebar::opensSidebar): Method to check if the
current sidebar has a valid path, enabling it to be opened
(WebKit::WebExtensionSidebar::isDefaultSidebar const): Method to check
if this sidebar object is the default sidebar object
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebarCocoa.mm: Added.
(WebKit::getDefaultSidebarTitleFromExtension): Utility function to
retrieve the manifest-specified default sidebar title for an extension
(WebKit::getDefaultSidebarPathFromExtension): Utility function to
retrieve the manifest-specified default sidebar path for an extension
(WebKit::WebExtensionSidebar::WebExtensionSidebar): Constructor for
WebExtensionSidebar which populates instance variables and performs a
simple sanity check on the provided window and tab values
(WebKit::WebExtensionSidebar::extensionContext const): Getter for
this sidebar&apos;s context
(WebKit::WebExtensionSidebar::tab const): Getter for this
sidebar&apos;s tab
(WebKit::WebExtensionSidebar::window const): Getter for this
sidebar&apos;s window
(WebKit::WebExtensionSidebar::parent const): Getter for this
sidebar&apos;s parent. It will check the current context for a less-specific
sidebar override if present, otherwise returning the default sidebar. If
this sidebar is the default sidebar, it will return std::nullopt (since
the default sidebar has no parent).
(WebKit::WebExtensionSidebar::propertiesDidChange): Function to notify
the browser through the yet-unimplemented sidebar delegate methods that
the properties of this sidebar have changed.
(WebKit::WebExtensionSidebar::icon): Getter for the icon override of this
sidebar, or this sidebar&apos;s parent if there is no override set.
(WebKit::WebExtensionSidebar::setIconsDictionary): Setter for this
sidebar&apos;s icon override
(WebKit::WebExtensionSidebar::title const): Getter for the title
override of this sidebar, or this sidebar&apos;s parent if there is no
override set.
(WebKit::WebExtensionSidebar::setTitle): Setter for this
sidebar&apos;s title override.
(WebKit::WebExtensionSidebar::isEnabled const): Getter to determine if
this specific sidebar is enabled.
(WebKit::WebExtensionSidebar::setEnabled): Setter to set whether this
specific sidebar is enabled.
(WebKit::WebExtensionSidebar::canProgrammaticallyOpenSidebar const):
Function to check whether it is possible to programmatically open this
sidebar. It ensures that this extension&apos;s context is present, and that
the context contains a valid extension controller object
(WebKit::WebExtensionSidebar::openSidebarWhenReady): Currently
unimplemented method to open this sidebar when the delegate is ready to
do so.
(WebKit::WebExtensionSidebar::canProgrammaticallyCloseSidebar const):
Functino to check whether it is possible to programmatically close this
sidebar. It ensures that this extension&apos;s context is present, and that
the context contains a valid extension controller object.
(WebKit::WebExtensionSidebar::closeSidebarWhenReady): Currently
unimplemented method to close this sidebar when the delegate is ready to
do so.
(WebKit::WebExtensionSidebar::sidebarPath const): Getter for this
sidebar&apos;s path override, or its parent&apos;s if there is no override set.
(WebKit::WebExtensionSidebar::setSidebarPath): Setter for this
particualr sidebar&apos;s path override.
(WebKit::WebExtensionSidebar::webView): Getter for this sidebar&apos;s
WebView.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h: Add member
  variables to service sidebar management methods, add sidebar
management method declarations.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj: Add
  WebExtensionSidebar.h and WebExtensionSidebar.mm to the project.

Canonical link: <a href="https://commits.webkit.org/281881@main">https://commits.webkit.org/281881@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/53f445f9bfff502dc4b2258cafcd46dbb979493e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/61315 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/40676 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/13896 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/65265 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/11864 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/48354 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/12139 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/49539 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/8242 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/63777 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/37814 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/53112 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/30377 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/34488 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/10777 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/56303 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/10641 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/66996 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/5262 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/10429 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/56911 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/5285 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/53078 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/57118 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13668 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/4344 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/36480 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/37563 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/38657 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/37307 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebar.h:
```diff
@@ -0,0 +1,121 @@
+
+    bool operator==(const WebExtensionSidebar&) const;
+
+    std::optional<Ref<WebExtensionContext const>> extensionContext() const;
```
I would drop `const` from `Ref<WebExtensionContext const>`.

AriYoung00 replied:
I tried to write both const and non-const getters for context, window, and tab -- the non-const getter drops the const within ref. I think that since this method is a const method, and since context is a property of this class, we would need this getter to return with the interior const right?

Edit: Actually ignore this, removing the const getters as per below comment

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebar.h:
```diff
@@ -0,0 +1,121 @@
+
+    std::optional<Ref<WebExtensionContext const>> extensionContext() const;
+    std::optional<Ref<WebExtensionContext>> extensionContext();
+    std::optional<Ref<WebExtensionTab const>> tab() const;
```
Ditto on `const` in the `Ref`.

AriYoung00 replied:
For `tab` and `window` I moved `const` to the beginning of the signature (`const std::optional<...>`) since that's the signature of the member variables these methods get

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -2622,6 +2625,48 @@ static WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+        return *windowAction;
+
+    return std::nullopt;
+}
```
Shouldn't this return the default if there is no window specific sidebar?

AriYoung00 replied:
I was planning to add the fallback-to-default behavior in a method which is "further up" -- in my local working changes (planning to put up another PR for that as soon as this one is landed) I have a method called `getSidebarWithProperties` in `WebExtensionContextAPISidebar` (more or less mirroring the same method in `WebExtensionAction`) which will return the default sidebar if one isn't found for the given properties (can be found [here on my WIP branch](https://github.com/AriYoung00/WebKit/blob/eng/ari_young/sidebar-sidepanel-more-state-wip/Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPISidebarCocoa.mm#L44)).

xeenon replied:
I think you will still need this to return the default for the ObjC API case. But we can revisit this if needed later on.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -2622,6 +2625,48 @@ static WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+        return *tabAction;
+
+    return std::nullopt;
+}
```
Ditto on returning the default if no tab specific sidebar is defined.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -2622,6 +2625,48 @@ static WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+    return std::nullopt;
+}
+
+Ref<WebExtensionSidebar> WebExtensionContext::getOrCreateSidebar(WebExtensionWindow& window)
```
I think this will need to be `optional` since if the extension does not use sidebars there is nothing to return.

AriYoung00 replied:
If the extension doesn't use sidebars, wouldn't these methods never be called? Just since we deny access to the `sidebarAction` / `sidePanel` properties in `WebExtensionAPINamespace::isPropertyAllowed` based on what's present in the manifest.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -2622,6 +2625,48 @@ static WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+    }).iterator->value;
+}
+
+Ref<WebExtensionSidebar> WebExtensionContext::getOrCreateSidebar(WebExtensionTab& tab)
```
Ditto on `optional`.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebar.h:
```diff
@@ -0,0 +1,121 @@
+    std::optional<Ref<WebExtensionContext>> extensionContext();
+    std::optional<Ref<WebExtensionTab const>> tab() const;
+    std::optional<Ref<WebExtensionTab>> tab();
+    std::optional<Ref<WebExtensionWindow const>> window() const;
```
Ditto on `const` in the `Ref`.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebar.h:
```diff
@@ -0,0 +1,121 @@
+    std::optional<Ref<WebExtensionTab>> tab();
+    std::optional<Ref<WebExtensionWindow const>> window() const;
+    std::optional<Ref<WebExtensionWindow>> window();
+    std::optional<Ref<WebExtensionSidebar const>> parent() const;
```
Ditto on `const` in the `Ref`.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebar.h:
```diff
@@ -0,0 +1,121 @@
+
+    /// `title()` will return the overridden title of this sidebar, or the title of the first parent sidebar in which the title is set
+    String title() const;
+    void setTitleOverride(std::optional<String>);
```
I think `setTitle` is clear enough, especially since it takes an `optional` that resets it.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebar.h:
```diff
@@ -0,0 +1,121 @@
+
+    /// `sidebarPath()` will return the overriden path of this sidebar, or the path of the first parent sidebar in which the path is set
+    String sidebarPath() const;
+    void setSidebarPathOverride(std::optional<String>);
```
Ditto, drop `Override`.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebar.h:
```diff
@@ -0,0 +1,121 @@
+    String sidebarPath() const;
+    void setSidebarPathOverride(std::optional<String>);
+
+    WKWebView *sidebarWebView();
```
I think just `webView` is fine here. I used `popupWebView` since action is not just for "popups". But "sidebar" is redundant in a `WebExtensionSidebar` class.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebar.h:
```diff
@@ -0,0 +1,121 @@
+    WTF_MAKE_NONCOPYABLE(WebExtensionSidebar);
+
+public:
+    enum class IsDefaultSidebar { No, Yes };
```
Just `IsDefault` is fine, since this is scoped to `WebExtensionSidebar`.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebar.h:
```diff
@@ -0,0 +1,121 @@
+    bool m_isEnabled { false };
+    const IsDefaultSidebar m_isDefault { IsDefaultSidebar::No };
+
+    RetainPtr<WKWebView> m_sidebarWebView;
```
`m_webView`

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebarCocoa.mm:
```diff
@@ -0,0 +1,299 @@
+static NSString * const sidePanelManifestKey = @"side_panel";
+static NSString * const sidebarActionTitleKey = @"default_title";
+static NSString * const sidebarActionPathKey = @"default_panel";
+static NSString * const sidePanelPathKey = @"default_path";
```
All manifest parsing should happen in `WebExtension` where `WebExtensionSidebar` can ask for default values like we do for `WebExtensionAction` using `extension().actionPopupPath()`, etc. This keeps the manifest parsing centralized and able to change as new properties / versions are supported. That would abstract away the `sidebar_action` vs `side_panel` stuff and make `WebExtensionSidebar` simpler and more flexible down the road.

AriYoung00 replied:
Just moved all of this over to WebExtension -- how does all the parsing stuff look now?

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebarCocoa.mm:
```diff
@@ -0,0 +1,299 @@
+    if (![maybeDict isKindOfClass:NSDictionary.class])
+        return std::nullopt;
+    return (NSDictionary<NSString *, id> *) maybeDict;
+}
```
Could move to `CocoaHelpers` and be a generic template. Something like:

```suggestion
template<typename T>
std::optional<RetainPtr<T>> optional_objc_cast(id object)
{
    if (auto *result = dynamic_objc_cast<T>(object))
        return result;
    return std::nullopt;
}
```

AriYoung00 replied:
I ended up removing this in favor of existing nullable pointer methods for the parsing code I'm moving to `WebExtension` -- it feels like it fits in better with the other existing `WebExtension` methods than something based on `std::optional` would, and I can just perform the conversion to optional in `WebExtensionSidebarCocoa.cpp`.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebarCocoa.mm:
```diff
@@ -0,0 +1,299 @@
+    if (![maybeValue isKindOfClass:NSString.class])
+        return std::nullopt;
+    return (NSString *) maybeValue;
+}
```
Also could move to `CocoaHelpers` and use `objectForKey`. Could also consider making a `template` so it is not `String` specific.

```suggestion
std::optional<String> optionalStringForKey(RetainPtr<NSDictionary> dictionary, id key)
{
    if (auto *string = objectForKey<NSString>(dictionary, key.get()))
        return String(string);
    return std::nullopt;
}
```

AriYoung00 replied:
I also removed this one in favor of existing pointer-based methods in `WebExtension`

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebarCocoa.mm:
```diff
@@ -0,0 +1,299 @@
+{
+    // FIXME: <https://webkit.org/b/276833> implement this
+    return std::nullopt;
+}
```
I think a lot of this would move to `WebExtension` since this is manifest parsing. See other comment at the top.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebarCocoa.mm:
```diff
@@ -0,0 +1,299 @@
+    if (maybeNil)
+        return maybeNil;
+    return std::nullopt;
+}
```
Move to `CocoaHelpers`, maybe as `toOptional`. But also needs to use `RetainPtr` otherwise this will leak objects.

```suggestion
template<typename T>
static std::optional<RetainPtr<T>> toOptional(T *maybeNil)
{
    if (maybeNil)
        return maybeNil;
    return std::nullopt;
}
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebarCocoa.mm:
```diff
@@ -0,0 +1,299 @@
+
+WebExtensionSidebar::WebExtensionSidebar(WebExtensionContext& context, std::optional<Ref<WebExtensionTab>> tab, std::optional<Ref<WebExtensionWindow>> window, IsDefaultSidebar isDefault)
+    // for this API, we never want to override both window and tab -- can only do one at a time
+    // if we somehow (erroneously) get both, prefer tab (most specific)
```
This comment should be an `ASSERT` instead. Then just `m_window(window)`.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebarCocoa.mm:
```diff
@@ -0,0 +1,299 @@
+    }
+}
+
+std::optional<Ref<WebExtensionContext const>> WebExtensionSidebar::extensionContext() const
```
We don't get much from this being `optional` as you only use `value().get()` when it could be `RefPtr` instead.

AriYoung00 replied:
There are a couple places where I'm using some of the monadic optional operations on `extensionContext` -- most notably in `canProgrammaticallyOpenSidebar` and `canProgrammaticallyCloseSidebar`. When I address the FIXMEs in both of those methods, I think I will also add another chained call to `and_then` in each to check that the delegates respond to the appropriate selectors, so I think the final implementation might be cleaner if I retain the `optional`s -- would it be alright to keep this one as an `optional<Ref<..>>`?

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebarCocoa.mm:
```diff
@@ -0,0 +1,299 @@
+    return const_cast<WebExtensionSidebar const*>(this)->window().transform([](auto window) -> Ref<WebExtensionWindow> {
+        return const_cast<WebExtensionWindow&>(window.get());
+    });
+}
```
Should just have one `window()` and one `tab()` that is `std::optional<Ref<T>>`, no `const`.

`transform` and lambdas seem like overkill for these simple getters.

AriYoung00 replied:
Understood, I can definitely remove the `const`/non-`const` duplication for these

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebarCocoa.mm:
```diff
@@ -0,0 +1,299 @@
+        return std::nullopt;
+
+    return m_tab
+        .and_then([this](auto tab) { return m_context->getSidebar(*tab->window()); })
```
```suggestion
        .and_then([this](auto& tab) { return m_context->getSidebar(*tab->window()); })
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebarCocoa.mm:
```diff
@@ -0,0 +1,299 @@
+
+std::optional<Ref<WebExtensionSidebar>> WebExtensionSidebar::parent()
+{
+    return const_cast<WebExtensionSidebar const*>(this)->parent().transform([](auto parent) -> Ref<WebExtensionSidebar> {
```
```suggestion
    return const_cast<WebExtensionSidebar const*>(this)->parent().transform([](auto& parent) -> Ref<WebExtensionSidebar> {
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebarCocoa.mm:
```diff
@@ -0,0 +1,299 @@
+
+std::optional<Ref<WebExtensionTab const>> WebExtensionSidebar::tab() const
+{
+    return m_tab.transform([](auto tab) -> Ref<WebExtensionTab const> {
```
```suggestion
    return m_tab.transform([](auto& tab) -> Ref<WebExtensionTab const> {
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebarCocoa.mm:
```diff
@@ -0,0 +1,299 @@
+    return const_cast<WebExtensionSidebar const*>(this)->extensionContext().transform([](auto ctx) -> Ref<WebExtensionContext> {
+        return const_cast<WebExtensionContext&>(ctx.get());
+    });
+}
```
Also no need for two `extensionContext()` methods, just non-const is fine.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebarCocoa.mm:
```diff
@@ -0,0 +1,299 @@
+
+std::optional<Ref<WebExtensionTab>> WebExtensionSidebar::tab()
+{
+    return const_cast<WebExtensionSidebar const*>(this)->tab().transform([](auto tab) -> Ref<WebExtensionTab> {
```
```suggestion
    return const_cast<WebExtensionSidebar const*>(this)->tab().transform([](auto& tab) -> Ref<WebExtensionTab> {
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebarCocoa.mm:
```diff
@@ -0,0 +1,299 @@
+
+std::optional<Ref<WebExtensionWindow const>> WebExtensionSidebar::window() const
+{
+    return m_window.transform([](auto window) -> Ref<WebExtensionWindow const> {
```
```suggestion
    return m_window.transform([](auto& window) -> Ref<WebExtensionWindow const> {
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebarCocoa.mm:
```diff
@@ -0,0 +1,299 @@
+
+std::optional<Ref<WebExtensionWindow>> WebExtensionSidebar::window()
+{
+    return const_cast<WebExtensionSidebar const*>(this)->window().transform([](auto window) -> Ref<WebExtensionWindow> {
```
```suggestion
    return const_cast<WebExtensionSidebar const*>(this)->window().transform([](auto& window) -> Ref<WebExtensionWindow> {
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebarCocoa.mm:
```diff
@@ -0,0 +1,299 @@
+            return nilToNullopt(context.extension().bestImageInIconsDictionary(icons.get(), largestDim(size)));
+        })
+        .or_else([&] -> std::optional<CocoaImage *> {
+            return parent().transform([&](auto parent) { return parent.get().icon(size); });
```
```suggestion
        }).or_else([&] -> std::optional<RetainPtr<CocoaImage>> {
            return parent().transform([&](auto& parent) { return parent.get().icon(size); });
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebarCocoa.mm:
```diff
@@ -0,0 +1,299 @@
+        m_iconsOverride = std::nullopt;
+        return;
+    }
+    if (m_iconsOverride && [m_iconsOverride.value() isEqualToDictionary:icons])
```
```suggestion
    }

    if (m_iconsOverride && [m_iconsOverride.value() isEqualToDictionary:icons])
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebarCocoa.mm:
```diff
@@ -0,0 +1,299 @@
+String WebExtensionSidebar::title() const
+{
+    return m_titleOverride.value_or(
+        parent().transform([](const auto parent) { return parent.get().title(); }).value_or(fallbackTitle)
```
```suggestion
        parent().transform([](auto& parent) { return parent.get().title(); }).value_or(fallbackTitle)
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebarCocoa.mm:
```diff
@@ -0,0 +1,299 @@
+
+bool WebExtensionSidebar::canProgrammaticallyOpenSidebar() const
+{
+    return extensionContext().transform([](auto context) -> bool { return !!context.get().extensionController(); })
```
```suggestion
    return extensionContext().transform([](auto& context) -> bool { return !!context.get().extensionController(); })
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebarCocoa.mm:
```diff
@@ -0,0 +1,299 @@
+
+bool WebExtensionSidebar::canProgrammaticallyCloseSidebar() const
+{
+    return extensionContext().transform([](auto context) -> bool { return !!context.get().extensionController(); })
```
```suggestion
    return extensionContext().transform([](auto& context) -> bool { return !!context.get().extensionController(); })
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebarCocoa.mm:
```diff
@@ -0,0 +1,299 @@
+String WebExtensionSidebar::sidebarPath() const
+{
+    return m_sidebarPathOverride.value_or(
+        parent().transform([](const auto parent) { return parent.get().sidebarPath(); }).value_or(fallbackPath)
```
```suggestion
        parent().transform([](auto& parent) { return parent.get().sidebarPath(); }).value_or(fallbackPath)
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
```diff
@@ -437,6 +438,14 @@ class WebExtensionContext : public API::ObjectImpl<API::Object::Type::WebExtensi
+    std::optional<Ref<WebExtensionSidebar>> getSidebar(WebExtensionWindow const&);
+    std::optional<Ref<WebExtensionSidebar>> getSidebar(WebExtensionTab const&);
+    Ref<WebExtensionSidebar> getOrCreateSidebar(WebExtensionWindow&);
+    Ref<WebExtensionSidebar> getOrCreateSidebar(WebExtensionTab&);
```
See comment about these needing to be `optional` too.

xeenon commented on Source/WebKit/Platform/cocoa/CocoaHelpers.h:
```diff
@@ -93,6 +93,21 @@ inline bool boolForKey(NSDictionary *dictionary, id key, bool defaultValue)
+    return std::nullopt;
+}
+
+inline std::optional<String> toOptional(NSString *maybeNil)
```
Nice!

AriYoung00 replied:
Thanks!

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -1114,6 +1122,65 @@ static WKWebExtensionError toAPI(WebExtension::Error error)
+    // see "Examples" header -> "Side Panel" tab (doesn't mention `default_path` key elsewhere)
+    // sidebarAction documentation: https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/sidebar_action
+
+    auto sidebarActionDict = objectForKey<NSDictionary>(m_manifest, sidebarActionManifestKey);
```
```suggestion
    auto *sidebarActionDictionary = objectForKey<NSDictionary>(m_manifest, sidebarActionManifestKey);
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -1114,6 +1122,65 @@ static WKWebExtensionError toAPI(WebExtension::Error error)
+        return;
+    }
+
+    auto sidePanelDict = objectForKey<NSDictionary>(m_manifest, sidePanelManifestKey);
```
```suggestion
    auto *sidePanelDictionary = objectForKey<NSDictionary>(m_manifest, sidePanelManifestKey);
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -1114,6 +1122,65 @@ static WKWebExtensionError toAPI(WebExtension::Error error)
+        populateSidePanelProperties(sidePanelDict);
+}
+
+void WebExtension::populateSidebarActionProperties(RetainPtr<NSDictionary> sidebarActionDict)
```
```suggestion
void WebExtension::populateSidebarActionProperties(RetainPtr<NSDictionary> sidebarActionDictionary)
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -1114,6 +1122,65 @@ static WKWebExtensionError toAPI(WebExtension::Error error)
+    m_sidebarDocumentPath = objectForKey<NSString>(sidebarActionDict, sidebarActionPathManifestKey);
+}
+
+void WebExtension::populateSidePanelProperties(RetainPtr<NSDictionary> sidePanelDict)
```
```suggestion
void WebExtension::populateSidePanelProperties(RetainPtr<NSDictionary> sidePanelDictionary)
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -2622,6 +2625,46 @@ static WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+    return *m_defaultSidebar;
+}
+
+std::optional<Ref<WebExtensionSidebar>> WebExtensionContext::getSidebar(WebExtensionWindow const& window)
```
Still not sure about `const` here. Was it needed to keep a const method happy?

```suggestion
std::optional<Ref<WebExtensionSidebar>> WebExtensionContext::getSidebar(WebExtensionWindow& window)
```

AriYoung00 replied:
More or less -- since the `m_window` and `m_tab` members of `WebExtensionSidebar` are const-members, they can only be passed by const reference.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -2622,6 +2625,46 @@ static WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+    return std::nullopt;
+}
+
+std::optional<Ref<WebExtensionSidebar>> WebExtensionContext::getSidebar(WebExtensionTab const& tab)
```
Ditto.

```suggestion
std::optional<Ref<WebExtensionSidebar>> WebExtensionContext::getSidebar(WebExtensionTab& tab)
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebarCocoa.mm:
```diff
@@ -0,0 +1,220 @@
+WebExtensionSidebar::WebExtensionSidebar(WebExtensionContext& context, WebExtensionWindow& window) : WebExtensionSidebar(context, std::nullopt, window, IsDefault::No) { };
+
+WebExtensionSidebar::WebExtensionSidebar(WebExtensionContext& context, std::optional<Ref<WebExtensionTab>> tab, std::optional<Ref<WebExtensionWindow>> window, IsDefault isDefault)
+    : m_context(WeakRef(context)), m_tab(tab), m_window(window), m_isDefault(isDefault)
```
Is `WeakRef()` needed?

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebarCocoa.mm:
```diff
@@ -0,0 +1,220 @@
+    return m_tab;
+}
+
+
```
Remove extra line.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebarCocoa.mm:
```diff
@@ -0,0 +1,220 @@
+        return std::nullopt;
+
+    return m_tab
+        .and_then([this](auto const& tab) { return m_context->getSidebar(*tab->window()); })
```
```suggestion
    return m_tab.and_then([this](auto const& tab) { return m_context->getSidebar(*tab->window()); })
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebarCocoa.mm:
```diff
@@ -0,0 +1,220 @@
+    String path = m_sidebarPathOverride.value_or(
+        parent().transform([](auto const& parent) { return parent.get().sidebarPath(); }).value_or(fallbackPath)
+    );
+    return path;
```
```suggestion
    return m_sidebarPathOverride.value_or(
        parent().transform([](auto const& parent) { return parent.get().sidebarPath(); }).value_or(fallbackPath)
    );
```

AriYoung00 replied:
Whoops, this is left over from when I was resolving some typing issues this morning -- meant to remove it before pushing

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
```diff
@@ -437,6 +438,14 @@ class WebExtensionContext : public API::ObjectImpl<API::Object::Type::WebExtensi
     void performAction(WebExtensionTab*, UserTriggered = UserTriggered::No);
 
+#if ENABLE(WK_WEB_EXTENSIONS_SIDEBAR)
+    WebExtensionSidebar& defaultSidebar();
```
Likely will also be optional, but we can change later if needed.

AriYoung00 committed:
Add WebExtensionSidebar object to UIProcess.
https://webkit.org/b/277574
rdar://133073139

Reviewed by Timothy Hatcher.

This PR adds the WebExtensionSidebar class to UIProcess. The structure
of this class closely mirrors that of WebExtensionAction, except it
prefers to use std::optional over nullable pointers when possible. This
PR also adds several utility methods in WebExtensionContext which will
be used to manage the WebExtensionSidebar objects necessary to service
the sidePanel / sidebarAction APIs for each extensions that utilizes
them.

* Source/WebKit/Platform/cocoa/CocoaHelpers.h:
 (WebKit::toOptional): Add helper method to convert nulltable T * (where
T is some Cocoa type) to optional<RetainPtr<T>>, where the interior
RetainPtr is guaranteed to be non-null
* Source/WebKit/Shared/API/APIObject.h: Add WebExtensionSidebar to
  API::Object::Type enum
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::sidebarIcon): Getter for current extension's
default sidebar icon
(WebKit::WebExtension::sidebarDocumentPath): Getter for current
extension's default document path
(WebKit::WebExtension::sidebarTitle): Getter for current extension's
default sidebar title
(WebKit::WebExtension::populateSidebarPropertiesIfNeeded): Helper method
to populate sidebar-related properties if needed
(WebKit::WebExtension::populateSidebarActionProperties): Helper method
to populate sidebar-related properties when sidebarAction is detected
(WebKit::WebExtension::populateSidePanelProperties): Helper method to
populate sidebar-related properties when sidePanel is detected
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::unload): Set m_defaultSidebar to nullptr
on unload
(WebKit::WebExtensionContext::defaultSidebar): Getter for
m_defaultSidebar
(WebKit::WebExtensionContext::getSidebar): Getters for window- or
tab-specific sidebar objects
(WebKit::WebExtensionContext::getOrCreateSidebar): Functions to create
and get window- or tab-specific sidebar objects
* Source/WebKit/UIProcess/Extensions/WebExtensionSidebar.h: Added.
(WebKit::WebExtensionSidebar::create): Canonical WebKit create method
(WebKit::WebExtensionSidebar::isOpen const): Getter for m_isOpen
(WebKit::WebExtensionSidebar::opensSidebar): Method to check if the
current sidebar has a valid path, enabling it to be opened
(WebKit::WebExtensionSidebar::isDefaultSidebar const): Method to check
if this sidebar object is the default sidebar object
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebarCocoa.mm: Added.
(WebKit::getDefaultSidebarTitleFromExtension): Utility function to
retrieve the manifest-specified default sidebar title for an extension
(WebKit::getDefaultSidebarPathFromExtension): Utility function to
retrieve the manifest-specified default sidebar path for an extension
(WebKit::WebExtensionSidebar::WebExtensionSidebar): Constructor for
WebExtensionSidebar which populates instance variables and performs a
simple sanity check on the provided window and tab values
(WebKit::WebExtensionSidebar::extensionContext const): Getter for
this sidebar's context
(WebKit::WebExtensionSidebar::tab const): Getter for this
sidebar's tab
(WebKit::WebExtensionSidebar::window const): Getter for this
sidebar's window
(WebKit::WebExtensionSidebar::parent const): Getter for this
sidebar's parent. It will check the current context for a less-specific
sidebar override if present, otherwise returning the default sidebar. If
this sidebar is the default sidebar, it will return std::nullopt (since
the default sidebar has no parent).
(WebKit::WebExtensionSidebar::propertiesDidChange): Function to notify
the browser through the yet-unimplemented sidebar delegate methods that
the properties of this sidebar have changed.
(WebKit::WebExtensionSidebar::icon): Getter for the icon override of this
sidebar, or this sidebar's parent if there is no override set.
(WebKit::WebExtensionSidebar::setIconsDictionary): Setter for this
sidebar's icon override
(WebKit::WebExtensionSidebar::title const): Getter for the title
override of this sidebar, or this sidebar's parent if there is no
override set.
(WebKit::WebExtensionSidebar::setTitle): Setter for this
sidebar's title override.
(WebKit::WebExtensionSidebar::isEnabled const): Getter to determine if
this specific sidebar is enabled.
(WebKit::WebExtensionSidebar::setEnabled): Setter to set whether this
specific sidebar is enabled.
(WebKit::WebExtensionSidebar::canProgrammaticallyOpenSidebar const):
Function to check whether it is possible to programmatically open this
sidebar. It ensures that this extension's context is present, and that
the context contains a valid extension controller object
(WebKit::WebExtensionSidebar::openSidebarWhenReady): Currently
unimplemented method to open this sidebar when the delegate is ready to
do so.
(WebKit::WebExtensionSidebar::canProgrammaticallyCloseSidebar const):
Functino to check whether it is possible to programmatically close this
sidebar. It ensures that this extension's context is present, and that
the context contains a valid extension controller object.
(WebKit::WebExtensionSidebar::closeSidebarWhenReady): Currently
unimplemented method to close this sidebar when the delegate is ready to
do so.
(WebKit::WebExtensionSidebar::sidebarPath const): Getter for this
sidebar's path override, or its parent's if there is no override set.
(WebKit::WebExtensionSidebar::setSidebarPath): Setter for this
particualr sidebar's path override.
(WebKit::WebExtensionSidebar::webView): Getter for this sidebar's
WebView.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h: Add member
  variables to service sidebar management methods, add sidebar
management method declarations.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj: Add
  WebExtensionSidebar.h and WebExtensionSidebar.mm to the project.

Canonical link: https://commits.webkit.org/281881@main
